Woo: Add product endpoint

0nko opened the pull request:
This PR is the result of a successful trial. It adds the REST client and DB support for the add-product endpoint.

It's already been reviewed but I'd like to get a final quick look before we merge it to `develop`.

**The original description:**

## RESOURCES
[The api definition](https://woocommerce.github.io/woocommerce-rest-api-docs/#create-a-product)

## CHANGES / DESCRIPTION

This PR will contain the work to allow a remote communication for the **creation of a new product**. This chunk of work meant:
- add new product actions in `WCProductAction`
- handle of new actions and new event `OnProductCreated` in `WCProductStore`
- add new product function in `ProductRestClient`
- added tests to the new request in **APITesting_WCProduct**, **ReleaseStack_WCProductTest** and **WCProductStoreTest.kt**

Following the `README`, I added the initial action in the `WCProductAction`; I called them:
- **ADD_NEW_PRODUCT** 
- **ADDED_NEW_PRODUCT**

In the `WCProductStore`, the `onAction` method will be listening to these actions that, upon trigger will behave as such:

**1)** `ADD_NEW_PRODUCT`
Will call the `createNewProduct` of the `ProductRestClient` that will make the remote connection and respond with success/error, mapping the response model accordingly. This done, it will trigger the correct action for completion, with the data;

**2)** `ADDED_NEW_PRODUCT`
Will call the `handleAddNewProduct`, from a previous completion of a request, and create the success/error scenarios accordingly, using the `OnProductCreated` for proper communication of the result.

This will get emitted to whoever is listening and trigger the UI updates (example app has this integration 👍 also, in screenshots bellow)

Let me know your thoughts on it 🙏 

## SCREENSHOTS
Product created
-
 <img width="315" src="https://user-images.githubusercontent.com/7058393/91237752-54c89000-e733-11ea-8c5b-29aa8fee387c.png"/>

New Product fetched
-
<img width="315" src="https://user-images.githubusercontent.com/7058393/91237755-55612680-e733-11ea-8357-79d4c54264d4.png"/><img width="315" src="https://user-images.githubusercontent.com/7058393/91237756-55612680-e733-11ea-9505-d8be39e1d45f.png"/>

Tests
-
<img width="415" alt="Screenshot 2020-08-25 at 23 56 17" src="https://user-images.githubusercontent.com/7058393/91237742-51350900-e733-11ea-98e3-519ad3d2ee62.png">

Commit:
Added the new add product actions. Added the respective payload classes. Added the respective payload handler. Added the createNewProduct rest client function. Added tests to the client

Commit:
Added the example UI on top of the update product one to allow creation of a new product

Commit:
Some tidy up of imports and some naming improvements

Commit:
Updated some naming as per PR comments

0nko committed:
Merge pull request #1676 from GuilhermeMota93/feat/2707-add-product

Feat/2707 add product

nbradbury commented:
In the example app, product width and height are duplicated and are sent to the backend as weight and length.

![Screenshot_1599662538](https://user-images.githubusercontent.com/3903757/92614384-fbf80c00-f289-11ea-9e2f-d87d014c9a9c.png)

nbradbury commented:
@0nko In the example app, many of the settings (product type, status, manage stock, etc) are disabled. Is this intentional?

nbradbury commented:
Also in the example app, I added a tag and a category but neither were sent to the backend.

![Screenshot_1599663543](https://user-images.githubusercontent.com/3903757/92615868-93119380-f28b-11ea-821d-d773d19543da.png)

<img width="618" alt="Screen Shot 2020-09-09 at 10 58 25 AM" src="https://user-images.githubusercontent.com/3903757/92615882-96a51a80-f28b-11ea-9dce-c2a49ee2f381.png">

nbradbury commented on example/src/androidTest/java/org/wordpress/android/fluxc/release/ReleaseStack_WCProductTest.kt:
```diff
@@ -891,4 +929,23 @@ class ReleaseStack_WCProductTest : ReleaseStack_WCBase() {
+        lastAddNewProductEvent = event
+
+        when (event.causeOfChange) {
+            WCProductAction.ADD_PRODUCT -> {
```
This test is failing for me because `event.causeOfChange` is `ADDED_PRODUCT` rather than `ADD_PRODUCT`.

<img width="697" alt="Screen Shot 2020-09-09 at 11 21 17 AM" src="https://user-images.githubusercontent.com/3903757/92619027-01a42080-f28f-11ea-8afe-c88cc607cd97.png">

0nko committed:
Remove duplicate code

0nko committed:
Fix the length & weight labels

0nko committed:
Initialize a new product model and enable all buttons

0nko committed:
FIx the failing test

0nko commented:
Thanks for the testing, @nbradbury! 

Seems like the field naming issues were already there and the other problems were related to initialization. I've addressed all issues, ready for another round.

nbradbury commented:
In the example app, if I tap the "Select Categories" or "Select Tags" buttons, any information I previously typed into the Add Product screen is erased when I return to it.

nbradbury commented:
I typed in a category and a tag, but neither is sent to the backend.

![Screenshot_1600079938](https://user-images.githubusercontent.com/3903757/93076659-adfe5200-f655-11ea-9d8a-1d72dbd9b90c.png)

0nko committed:
Update the initialization logic so the data is preserved when leaving the product update screen

0nko commented:
> In the example app, if I tap the "Select Categories" or "Select Tags" buttons, any information I previously typed into the Add Product screen is erased when I return to it.

This seems to be a problem if you have Don't keep activities enabled. Seems like an existing problem with the Update flow, as well. I updated the re-initialization logic so this should be fixed now.

> I typed in a category and a tag, but neither is sent to the backend.

That's not how it works. First, you need to fetch the Tags/Categories on the previous screen. Then you need to tap on the Tags/Categories button, select items from the list and tap on the Done button.

nbradbury commented:
@0nko Changes look good, but there's a failing test in `DateUtilsTest`. It doesn't look like this PR made any changes to that, and the test passes locally, so I'm not sure why it's failing.

0nko committed:
Attempt at fixing a failing DateTimeUtil unit tests

(cherry picked from commit 7020f4926b47c5799f17cfba5bea5783e76b15fc)

nbradbury reviewed (approved):
Looks good! :shipit:
